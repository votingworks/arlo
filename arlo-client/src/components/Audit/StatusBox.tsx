import React from 'react'
import { useParams } from 'react-router-dom'
import styled from 'styled-components'
import { Callout, H4 } from '@blueprintjs/core'
import { toast } from 'react-toastify'
import useRoundsAuditAdmin, { IRound } from './useRoundsAuditAdmin'
import useJurisdictions, {
  IJurisdiction,
  FileProcessingStatus,
  JurisdictionRoundStatus,
} from './useJurisdictions'
import FormButton from '../Atoms/Form/FormButton'
import { api } from '../utilities'
import { Inner } from '../Atoms/Wrapper'
import { IAuditSettings, IContest } from '../../types'
import useAuditSettings from './useAuditSettings'
import useContests from './useContests'

const Wrapper = styled(Callout)`
  display: flex;
  padding: 30px 0;
  .text {
    flex-grow: 1;
    p {
      margin-bottom: 0;
    }
  }
`
enum Button {
  START_NEXT_ROUND,
  DOWNLOAD_REPORT,
}

const createRound = async (electionId: string, roundNum: number) => {
  try {
    api(`/election/${electionId}/round`, {
      method: 'POST',
      body: JSON.stringify({
        roundNum,
      }),
      headers: {
        'Content-Type': 'application/json',
      },
    })
  } catch (err) {
    toast.error(err.message)
  }
}

const downloadAuditReport = (electionId: string) => {
  window.open(`/election/${electionId}/audit/report`)
}

export const isSetupComplete = (
  jurisdictions: IJurisdiction[],
  contests: IContest[],
  auditSettings: IAuditSettings
): boolean =>
  jurisdictions.length > 0 &&
  contests.some(c => c.isTargeted) &&
  Object.entries(auditSettings).every(([, v]) => v !== null)

const statusContent = (
  rounds: IRound[],
  jurisdictions: IJurisdiction[],
  contests: IContest[],
  auditSettings: IAuditSettings
): {
  headline: string
  details: string[]
  button: Button | null
} => {
  // Audit setup
  if (rounds.length === 0) {
    const details = [
      isSetupComplete(jurisdictions, contests, auditSettings)
        ? 'Audit setup is complete.'
        : 'Audit setup is not complete.',
    ]
    if (jurisdictions.length > 0) {
      const numUploaded = jurisdictions.filter(
        ({ ballotManifest: { processing } }) =>
          processing && processing.status === FileProcessingStatus.PROCESSED
      ).length
      details.push(
        `${numUploaded} of ${jurisdictions.length}` +
          ' jurisdictions have completed file uploads.'
      )
    }
    return {
      headline: 'The audit has not started.',
      details,
      button: null,
    }
  }

  const { roundNum, endedAt, isAuditComplete } = rounds[rounds.length - 1]

  // Round in progress
  if (!endedAt) {
    const numCompleted = jurisdictions.filter(
      ({ currentRoundStatus }) =>
        currentRoundStatus &&
        currentRoundStatus.status === JurisdictionRoundStatus.COMPLETE
    ).length
    return {
      headline: `Round ${roundNum} of the audit is in progress`,
      details: [
        `${numCompleted} of ${jurisdictions.length} jurisdictions` +
          ` have completed Round ${roundNum}`,
      ],
      button: Button.START_NEXT_ROUND,
    }
  }

  // Round complete, need another round
  if (!isAuditComplete) {
    return {
      headline: `Round ${roundNum} of the audit is complete - another round is needed`,
      details: [`When you are ready, start Round ${roundNum + 1}`],
      button: Button.DOWNLOAD_REPORT,
    }
  }

  // Round complete, audit complete
  return {
    headline: 'Congratulations - the audit is complete!',
    details: [],
    button: null,
  }
}

const StatusBox: React.FC = () => {
  const { electionId } = useParams<{ electionId: string }>()
  const rounds = useRoundsAuditAdmin(electionId)
  const jurisdictions = useJurisdictions(electionId)
  const [contests] = useContests(electionId)
  const [auditSettings] = useAuditSettings(electionId)

  if (!rounds) return null // Still loading

  const { headline, details, button } = statusContent(
    rounds,
    jurisdictions,
    contests,
    auditSettings
  )

  const buttonElement = (() => {
    switch (button) {
      case Button.START_NEXT_ROUND: {
        const { roundNum } = rounds[rounds.length - 1]
        return (
          <FormButton
            intent="success"
            onClick={() => createRound(electionId, roundNum + 1)}
          >
            Start Round {roundNum + 1}
          </FormButton>
        )
      }
      case Button.DOWNLOAD_REPORT:
        return (
          <FormButton
            intent="success"
            onClick={e => {
              e.preventDefault()
              downloadAuditReport(electionId)
            }}
          >
            Download Audit Report
          </FormButton>
        )
      default:
        return null
    }
  })()

  return (
    <Wrapper>
      <Inner>
        <div className="text">
          <H4>{headline}</H4>
          {details.map(detail => (
            <p key={detail}>{detail}</p>
          ))}
        </div>
        <div>{buttonElement}</div>
      </Inner>
    </Wrapper>
  )
}

export default StatusBox
