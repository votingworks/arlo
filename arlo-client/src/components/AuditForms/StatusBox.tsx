import React from 'react'
import styled from 'styled-components'
import { Callout, H4 } from '@blueprintjs/core'
import { IAudit } from '../../types'
import FormButton from '../Form/FormButton'

const Wrapper = styled(Callout)`
  .details {
    width: 50%;
  }
  .wrapper {
    margin: 10px 0 10px 0;
    text-align: right;
  }
`

const generateStatuses = (
  audit: IAudit,
  launched: boolean,
  started: boolean,
  complete: boolean
): [string, string | undefined] => {
  const STATUSES = {
    audit: [
      'The audit has not launched.',
      'The audit has not started.',
      `Round ${audit.rounds.length} of the audit is in progress.`,
      'The audit is complete.',
    ],
    setup: [
      'Audit setup is not complete.',
      'Audit setup is complete.',
      undefined,
    ],
  }

  if (complete) {
    // all jurisdictions have completed the audit
    return [STATUSES.audit[3], STATUSES.setup[2]]
  } else if (launched && started) {
    // rounds have started
    return [STATUSES.audit[2], STATUSES.setup[2]]
  } else if (launched && !started) {
    // after setup before first round
    return [STATUSES.audit[1], STATUSES.setup[1]]
  }
  /* istanbul ignore next */
  return [STATUSES.audit[0], STATUSES.setup[0]]
}

interface IProps {
  audit: IAudit
  electionId: string
  launched: boolean
  started: boolean
}

const StatusBox = ({ audit, electionId, launched, started }: IProps) => {
  const downloadAuditReport = async (e: React.FormEvent) => {
    e.preventDefault()
    window.open(`/election/${electionId}/audit/report`)
  }

  const complete = audit.rounds.length
    ? audit.rounds[audit.rounds.length - 1].contests.every(
        c => c.endMeasurements.isComplete
      )
    : false

  const [auditStatus, setupStatus] = generateStatuses(
    audit,
    launched,
    started,
    complete
  )

  return (
    <Wrapper>
      <H4>{auditStatus}</H4>
      <div className="details">{setupStatus && <p>{setupStatus}</p>}</div>
      <div className="wrapper">
        <FormButton
          disabled={!complete}
          intent={complete ? 'success' : 'none'}
          onClick={downloadAuditReport}
        >
          Download Audit Reports {!complete && '(Incomplete)'}
        </FormButton>
      </div>
    </Wrapper>
  )
}

export default StatusBox
