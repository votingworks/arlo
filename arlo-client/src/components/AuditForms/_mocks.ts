import { Audit } from '../../types'

export const mockAudit: Audit = {
  name: 'Primary 2019',
  riskLimit: 10,
  randomSeed: 123123123,

  contests: [
    {
      id: 'contest-1',
      name: 'Contest 1',

      choices: [
        {
          id: 'candidate-1',
          name: 'Candidate 1',
          numVotes: 42,
        },
        {
          id: 'candidate-2',
          name: 'Candidate 2',
          numVotes: 72,
        },
      ],

      totalBallotsCast: 4200,
    },
  ],

  jurisdictions: [
    {
      id: 'adams-county',
      name: 'Adams County',
      contests: ['contest-1'],
      auditBoards: [
        {
          id: 'audit-board-1',
          name: 'Audit Board #1',
          members: [],
        },
        {
          id: 'audit-board-2',
          name: 'Audit Board #2',
          members: [],
        },
      ],
      ballotManifest: {
        filename: 'Adams_County_Manifest.csv',
        numBallots: 123456,
        numBatches: 560,
        uploadedAt: '2019-06-17 11:45:00',
      },
    },
  ],

  rounds: [
    {
      startedAt: '2019-06-17 11:45:00',
      endedAt: '2019-06-17 11:55:00',
      contests: [
        {
          id: 'contest-1',
          endMeasurements: {
            pvalue: 0.085,
            isComplete: false,
          },
          results: {
            'candidate-1': 55,
            'candidate-2': 35,
          },
          sampleSize: 25,
        },
      ],
      jurisdictions: {
        'adams-county': {
          numBallots: 15,
        },
      },
    },
  ],
}

export const statusStates: Audit[] = [
  {
    name: '',
    riskLimit: '',
    randomSeed: '',
    contests: [],
    jurisdictions: [],
    rounds: [],
  },
  {
    contests: [
      {
        choices: [
          {
            id: 'choice-1',
            name: 'choice one',
            numVotes: 792,
          },
          {
            id: 'choice-2',
            name: 'choice two',
            numVotes: 1325,
          },
        ],
        id: 'contest-1',
        name: 'contest name',
        totalBallotsCast: 2123,
        sampleSizeOptions: [
          { size: 269, type: 'ASN' },
          { size: 379, prob: 0.8 },
          { size: 78 },
        ],
      },
    ],
    jurisdictions: [],
    rounds: [],
    name: 'contest name',
    randomSeed: '123456789',
    riskLimit: 1,
  },
  {
    contests: [
      {
        choices: [
          {
            id: 'choice-1',
            name: 'choice one',
            numVotes: 792,
          },
          {
            id: 'choice-2',
            name: 'choice two',
            numVotes: 1325,
          },
        ],
        id: 'contest-1',
        name: 'contest name',
        totalBallotsCast: 2123,
        sampleSizeOptions: [
          { size: 269, type: 'ASN' },
          { size: 379, prob: 0.8 },
          { size: 78 },
        ],
      },
    ],
    jurisdictions: [
      {
        auditBoards: [
          {
            id: 'audit-board-1',
            name: 'Audit Board #1',
            members: [],
          },
        ],
        ballotManifest: {
          filename: null,
          numBallots: null,
          numBatches: null,
          uploadedAt: null,
        },
        contests: ['contest-1'],
        id: 'jurisdiction-1',
        name: 'Jurisdiction 1',
      },
    ],
    rounds: [],
    name: 'contest name',
    randomSeed: '123456789',
    riskLimit: 1,
  },
  {
    contests: [
      {
        choices: [
          {
            id: 'choice-1',
            name: 'choice one',
            numVotes: 792,
          },
          {
            id: 'choice-2',
            name: 'choice two',
            numVotes: 1325,
          },
        ],
        id: 'contest-1',
        name: 'contest name',
        totalBallotsCast: 2123,
        sampleSizeOptions: [
          { size: 269, type: 'ASN' },
          { size: 379, prob: 0.8 },
          { size: 78 },
        ],
      },
    ],
    jurisdictions: [
      {
        auditBoards: [
          {
            id: 'audit-board-1',
            name: 'Audit Board #1',
            members: [],
          },
        ],
        ballotManifest: {
          filename: 'Ballot Manifest May 2019 Election - WYNADOTTE.csv',
          numBallots: 2117,
          numBatches: 10,
          uploadedAt: 'Thu, 18 Jul 2019 16:34:07 GMT',
        },
        contests: ['contest-1'],
        id: 'jurisdiction-1',
        name: 'Jurisdiction 1',
      },
    ],
    rounds: [
      {
        contests: [
          {
            endMeasurements: {
              isComplete: null,
              pvalue: null,
            },
            id: 'contest-1',
            results: {},
            sampleSize: 379,
          },
        ],
        endedAt: null,
        startedAt: 'Thu, 18 Jul 2019 16:34:07 GMT',
      },
    ],
    name: 'contest name',
    randomSeed: '123456789',
    riskLimit: 1,
  },
  {
    contests: [
      {
        choices: [
          {
            id: 'choice-1',
            name: 'choice one',
            numVotes: 792,
          },
          {
            id: 'choice-2',
            name: 'choice two',
            numVotes: 1325,
          },
        ],
        id: 'contest-1',
        name: 'contest name',
        totalBallotsCast: 2123,
        sampleSizeOptions: [
          { size: 269, type: 'ASN' },
          { size: 379, prob: 0.8 },
          { size: 78 },
        ],
      },
    ],
    jurisdictions: [
      {
        auditBoards: [
          {
            id: 'audit-board-1',
            name: 'Audit Board #1',
            members: [],
          },
        ],
        ballotManifest: {
          filename: 'Ballot Manifest May 2019 Election - WYNADOTTE.csv',
          numBallots: 2117,
          numBatches: 10,
          uploadedAt: 'Thu, 18 Jul 2019 16:34:07 GMT',
        },
        contests: ['contest-1'],
        id: 'jurisdiction-1',
        name: 'Jurisdiction 1',
      },
    ],
    rounds: [
      {
        contests: [
          {
            endMeasurements: {
              isComplete: true,
              pvalue: 0.00020431431380638307,
            },
            id: 'contest-1',
            results: {
              'choice-1': 100,
              'choice-2': 167,
            },
            sampleSize: 379,
          },
        ],
        endedAt: 'Thu, 18 Jul 2019 16:59:34 GMT',
        startedAt: 'Thu, 18 Jul 2019 16:34:07 GMT',
      },
    ],
    name: 'contest name',
    randomSeed: '123456789',
    riskLimit: 1,
  },
]

export default mockAudit
