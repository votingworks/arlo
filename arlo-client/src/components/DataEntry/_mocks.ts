import { IAuditBoard, IBallot, BallotStatus, Interpretation } from '../../types'

export const contest = {
  choices: [
    {
      id: 'choice-1',
      name: 'choice one',
      numVotes: 792,
    },
    {
      id: 'choice-2',
      name: 'choice two',
      numVotes: 1325,
    },
  ],
  id: 'contest-1',
  name: 'contest name',
  numWinners: '1',
  votesAllowed: '1',
  totalBallotsCast: '2123',
  isTargeted: true,
  jurisdictionIds: [],
}

export const doneDummyBallots: { ballots: IBallot[] } = {
  ballots: [
    {
      batch: {
        name: '0003-04-Precinct 13 (Jonesboro Fire Department)',
        id: 'batch-id-1',
        tabulator: '11',
      },
      position: 313,
      status: BallotStatus.AUDITED,
      interpretations: [],
      timesSampled: 1,
    },
    {
      batch: {
        name: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        id: 'batch-id-1',
        tabulator: '11',
      },
      position: 2112,
      status: BallotStatus.AUDITED,
      interpretations: [],
      timesSampled: 1,
    },
    {
      batch: {
        name: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        id: 'batch-id-1',
        tabulator: '11',
      },
      position: 1789,
      status: BallotStatus.AUDITED,
      interpretations: [],
      timesSampled: 1,
    },
  ],
}

export const dummyBallots: { ballots: IBallot[] } = {
  ballots: [
    {
      batch: {
        name: '0003-04-Precinct 13 (Jonesboro Fire Department)',
        id: 'batch-id-1',
        tabulator: null,
      },
      position: 313,
      status: BallotStatus.AUDITED,
      interpretations: [
        {
          contestId: contest.id,
          interpretation: Interpretation.VOTE,
          choiceId: contest.choices[0].id,
          comment: 'Good ballot',
        },
      ],
      timesSampled: 1,
    },
    {
      batch: {
        name: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        id: 'batch-id-1',
        tabulator: '11',
      },
      position: 2112,
      status: BallotStatus.NOT_AUDITED,
      interpretations: [],
      timesSampled: 1,
    },
    {
      batch: {
        name: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        id: 'batch-id-1',
        tabulator: '11',
      },
      position: 1789,
      status: BallotStatus.AUDITED,
      interpretations: [],
      timesSampled: 1,
    },
    {
      batch: {
        name: '0003-04-Precinct 13 (Jonesboro Fire Department)',
        id: 'batch-id-2',
        tabulator: '11',
      },
      position: 313,
      status: BallotStatus.AUDITED,
      interpretations: [],
      timesSampled: 1,
    },
    {
      batch: {
        name: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        id: 'batch-id-2',
        tabulator: '11',
      },
      position: 2112,
      status: BallotStatus.NOT_AUDITED,
      interpretations: [],
      timesSampled: 1,
    },
    {
      batch: {
        name: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        id: 'batch-id-2',
        tabulator: '11',
      },
      position: 1789,
      status: BallotStatus.AUDITED,
      interpretations: [],
      timesSampled: 1,
    },
    {
      batch: {
        name: '0003-04-Precinct 13 (Jonesboro Fire Department)',
        id: 'batch-id-3',
        tabulator: '11',
      },
      position: 313,
      status: BallotStatus.AUDITED,
      interpretations: [],
      timesSampled: 1,
    },
    {
      batch: {
        name: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        id: 'batch-id-3',
        tabulator: '11',
      },
      position: 2112,
      status: BallotStatus.NOT_AUDITED,
      interpretations: [],
      timesSampled: 1,
    },
    {
      batch: {
        name: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        id: 'batch-id-3',
        tabulator: '11',
      },
      position: 1789,
      status: BallotStatus.AUDITED,
      interpretations: [],
      timesSampled: 1,
    },
    {
      batch: {
        name: '0003-04-Precinct 13 (Jonesboro Fire Department)',
        id: 'batch-id-4',
        tabulator: '11',
      },
      position: 313,
      status: BallotStatus.AUDITED,
      interpretations: [],
      timesSampled: 1,
    },
    {
      batch: {
        name: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        id: 'batch-id-4',
        tabulator: '11',
      },
      position: 2112,
      status: BallotStatus.NOT_AUDITED,
      interpretations: [],
      timesSampled: 1,
    },
    {
      batch: {
        name: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        id: 'batch-id-4',
        tabulator: '11',
      },
      position: 1789,
      status: BallotStatus.AUDITED,
      interpretations: [],
      timesSampled: 1,
    },
    {
      batch: {
        name: '0003-04-Precinct 13 (Jonesboro Fire Department)',
        id: 'batch-id-5',
        tabulator: '11',
      },
      position: 313,
      status: BallotStatus.AUDITED,
      interpretations: [],
      timesSampled: 1,
    },
    {
      batch: {
        name: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        id: 'batch-id-5',
        tabulator: '11',
      },
      position: 2112,
      status: BallotStatus.NOT_AUDITED,
      interpretations: [],
      timesSampled: 1,
    },
    {
      batch: {
        name: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        id: 'batch-id-5',
        tabulator: '11',
      },
      position: 1789,
      status: BallotStatus.AUDITED,
      interpretations: [],
      timesSampled: 1,
    },
    {
      batch: {
        name: '0003-04-Precinct 13 (Jonesboro Fire Department)',
        id: 'batch-id-6',
        tabulator: '11',
      },
      position: 313,
      status: BallotStatus.AUDITED,
      interpretations: [],
      timesSampled: 1,
    },
    {
      batch: {
        name: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        id: 'batch-id-6',
        tabulator: '11',
      },
      position: 2112,
      status: BallotStatus.NOT_AUDITED,
      interpretations: [],
      timesSampled: 1,
    },
    {
      batch: {
        name: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        id: 'batch-id-6',
        tabulator: '11',
      },
      position: 1789,
      status: BallotStatus.AUDITED,
      interpretations: [],
      timesSampled: 1,
    },
    {
      batch: {
        name: '0003-04-Precinct 13 (Jonesboro Fire Department)',
        id: 'batch-id-7',
        tabulator: '11',
      },
      position: 313,
      status: BallotStatus.AUDITED,
      interpretations: [],
      timesSampled: 1,
    },
    {
      batch: {
        name: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        id: 'batch-id-7',
        tabulator: '11',
      },
      position: 2112,
      status: BallotStatus.NOT_AUDITED,
      interpretations: [],
      timesSampled: 1,
    },
    {
      batch: {
        name: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        id: 'batch-id-7',
        tabulator: '11',
      },
      position: 1789,
      status: BallotStatus.AUDITED,
      interpretations: [],
      timesSampled: 1,
    },
    {
      batch: {
        name: '0003-04-Precinct 13 (Jonesboro Fire Department)',
        id: 'batch-id-8',
        tabulator: '11',
      },
      position: 313,
      status: BallotStatus.AUDITED,
      interpretations: [],
      timesSampled: 1,
    },
    {
      batch: {
        name: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        id: 'batch-id-8',
        tabulator: '11',
      },
      position: 2112,
      status: BallotStatus.NOT_AUDITED,
      interpretations: [],
      timesSampled: 1,
    },
    {
      batch: {
        name: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        id: 'batch-id-8',
        tabulator: '11',
      },
      position: 1789,
      status: BallotStatus.AUDITED,
      interpretations: [],
      timesSampled: 1,
    },
    {
      batch: {
        name: '0003-04-Precinct 13 (Jonesboro Fire Department)',
        id: 'batch-id-9',
        tabulator: '11',
      },
      position: 313,
      status: BallotStatus.AUDITED,
      interpretations: [],
      timesSampled: 1,
    },
    {
      batch: {
        name: '0003-04-Precinct 19 (Jonesboro Fire Department)',
        id: 'batch-id-9',
        tabulator: '11',
      },
      position: 2112,
      status: BallotStatus.NOT_AUDITED,
      interpretations: [],
      timesSampled: 1,
    },
    {
      batch: {
        name: '0003-04-Precinct 29 (Jonesboro Fire Department)',
        id: 'batch-id-9',
        tabulator: '11',
      },
      position: 1789,
      status: BallotStatus.AUDITED,
      interpretations: [],
      timesSampled: 1,
    },
  ],
}

export const dummyBoard: IAuditBoard[] = [
  {
    id: '123',
    name: 'Audit Board #1',
    members: [],
  },
  {
    id: '123',
    name: 'Audit Board #1',
    members: [
      {
        name: 'John Doe',
        affiliation: '',
      },
      {
        name: 'Jane Doe',
        affiliation: 'LIB',
      },
    ],
  },
]

export default {
  dummyBoard,
  dummyBallots,
}
