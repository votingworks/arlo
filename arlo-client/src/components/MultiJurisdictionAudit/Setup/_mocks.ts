import { ElementType } from '../../../types'
import { setupStages } from './index'
import { ISidebarMenuItem } from '../../Atoms/Sidebar'

const relativeStages = (
  stage: ElementType<typeof setupStages>,
  state: ISidebarMenuItem['state'] = 'live'
): {
  prevStage: ISidebarMenuItem
  nextStage: ISidebarMenuItem
  menuItems: ISidebarMenuItem[]
} => {
  const prevTitle = setupStages[setupStages.indexOf(stage) - 1]
  const prevStage: ISidebarMenuItem = {
    title: prevTitle,
    active: false,
    activate: jest.fn(),
    state,
  }
  const nextTitle = setupStages[setupStages.indexOf(stage) + 1]
  const nextStage: ISidebarMenuItem = {
    title: nextTitle,
    active: false,
    activate: jest.fn(),
    state,
  }
  const menuItems = setupStages.map((s: ElementType<typeof setupStages>) => ({
    title: s,
    active: s === stage,
    activate: jest.fn(),
    state,
  }))
  return { prevStage, nextStage, menuItems }
}

export default relativeStages
