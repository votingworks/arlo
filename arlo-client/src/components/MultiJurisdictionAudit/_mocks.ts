import { IAuditSettings } from '../../types'
import {
  IJurisdiction,
  FileProcessingStatus,
  IFileInfo,
  JurisdictionRoundStatus,
} from './useJurisdictions' // uses IFileInfo instead of IBallotManifest and allows `file: null`
import { IRound } from './useRoundsJurisdictionAdmin' // has roundNum
import { IAuditBoard } from './useAuditBoards'

export const auditSettings: {
  [key in 'blank' | 'onlyState' | 'otherSettings' | 'all']: IAuditSettings
} = {
  blank: {
    state: null,
    electionName: null,
    online: null,
    randomSeed: null,
    riskLimit: null,
  },
  onlyState: {
    state: 'AL',
    electionName: null,
    online: null,
    randomSeed: null,
    riskLimit: null,
  },
  otherSettings: {
    state: null,
    electionName: 'Election Name',
    online: true,
    randomSeed: '12345',
    riskLimit: 10,
  },
  all: {
    state: 'AL',
    electionName: 'Election Name',
    online: true,
    randomSeed: '12345',
    riskLimit: 10,
  },
}

export const roundMocks: {
  [key in
    | 'empty'
    | 'singleIncomplete'
    | 'twoIncomplete'
    | 'singleComplete'
    | 'needAnother']: IRound[]
} = {
  empty: [],
  singleIncomplete: [
    {
      endedAt: null,
      roundNum: 1,
      isAuditComplete: false,
      startedAt: '2019-07-18T16:34:07.000Z',
      id: 'round-1',
    },
  ],
  twoIncomplete: [
    {
      endedAt: null,
      roundNum: 1,
      isAuditComplete: false,
      startedAt: '2019-07-18T16:34:07.000Z',
      id: 'round-1',
    },
    {
      endedAt: null,
      roundNum: 2,
      isAuditComplete: false,
      startedAt: '2019-07-18T16:34:07.000Z',
      id: 'round-2',
    },
  ],
  singleComplete: [
    {
      endedAt: 'a time most proper',
      roundNum: 1,
      isAuditComplete: true,
      startedAt: '2019-07-18T16:34:07.000Z',
      id: 'round-1',
    },
  ],
  needAnother: [
    {
      endedAt: 'a time most proper',
      roundNum: 1,
      isAuditComplete: false,
      startedAt: '2019-07-18T16:34:07.000Z',
      id: 'round-1',
    },
  ],
}

export const jurisdictionMocks: {
  [key in
    | 'empty'
    | 'twoUnprocessed'
    | 'oneUnprocessedOneProcessed'
    | 'twoProcessed'
    | 'oneComplete'
    | 'twoComplete']: IJurisdiction[]
} = {
  empty: [],
  twoUnprocessed: [
    {
      id: 'jurisdiction-id-1',
      name: 'Jurisdiction One',
      ballotManifest: { file: null, processing: null },
      currentRoundStatus: null,
    },
    {
      id: 'jurisdiction-id-2',
      name: 'Jurisdiction Two',
      ballotManifest: { file: null, processing: null },
      currentRoundStatus: null,
    },
  ],
  oneUnprocessedOneProcessed: [
    {
      id: 'jurisdiction-id-1',
      name: 'Jurisdiction One',
      ballotManifest: {
        file: null,
        processing: {
          status: FileProcessingStatus.PROCESSED,
          startedAt: 'sometime',
          completedAt: 'a different time',
          error: null,
        },
      },
      currentRoundStatus: null,
    },
    {
      id: 'jurisdiction-id-2',
      name: 'Jurisdiction Two',
      ballotManifest: { file: null, processing: null },
      currentRoundStatus: null,
    },
  ],
  twoProcessed: [
    {
      id: 'jurisdiction-id-1',
      name: 'Jurisdiction One',
      ballotManifest: {
        file: null,
        processing: {
          status: FileProcessingStatus.PROCESSED,
          startedAt: 'sometime',
          completedAt: 'a different time',
          error: null,
        },
      },
      currentRoundStatus: null,
    },
    {
      id: 'jurisdiction-id-2',
      name: 'Jurisdiction Two',
      ballotManifest: {
        file: null,
        processing: {
          status: FileProcessingStatus.PROCESSED,
          startedAt: 'sometime',
          completedAt: 'a different time',
          error: null,
        },
      },
      currentRoundStatus: null,
    },
  ],
  oneComplete: [
    {
      id: 'jurisdiction-id-1',
      name: 'Jurisdiction One',
      ballotManifest: {
        file: null,
        processing: {
          status: FileProcessingStatus.PROCESSED,
          startedAt: 'sometime',
          completedAt: 'a different time',
          error: null,
        },
      },
      currentRoundStatus: {
        status: JurisdictionRoundStatus.IN_PROGRESS,
        numBallotsAudited: 0,
        numBallotsSampled: 30,
      },
    },
    {
      id: 'jurisdiction-id-2',
      name: 'Jurisdiction Two',
      ballotManifest: {
        file: null,
        processing: {
          status: FileProcessingStatus.PROCESSED,
          startedAt: 'sometime',
          completedAt: 'a different time',
          error: null,
        },
      },
      currentRoundStatus: {
        status: JurisdictionRoundStatus.COMPLETE,
        numBallotsAudited: 30,
        numBallotsSampled: 30,
      },
    },
  ],
  twoComplete: [
    {
      id: 'jurisdiction-id-1',
      name: 'Jurisdiction One',
      ballotManifest: {
        file: null,
        processing: {
          status: FileProcessingStatus.PROCESSED,
          startedAt: 'sometime',
          completedAt: 'a different time',
          error: null,
        },
      },
      currentRoundStatus: {
        status: JurisdictionRoundStatus.COMPLETE,
        numBallotsAudited: 30,
        numBallotsSampled: 30,
      },
    },
    {
      id: 'jurisdiction-id-2',
      name: 'Jurisdiction Two',
      ballotManifest: {
        file: null,
        processing: {
          status: FileProcessingStatus.PROCESSED,
          startedAt: 'sometime',
          completedAt: 'a different time',
          error: null,
        },
      },
      currentRoundStatus: {
        status: JurisdictionRoundStatus.COMPLETE,
        numBallotsAudited: 30,
        numBallotsSampled: 30,
      },
    },
  ],
}

export const fileProcessingMocks: {
  [key in 'null' | 'processed']: IFileInfo['processing']
} = {
  null: null,
  processed: {
    status: FileProcessingStatus.PROCESSED,
    startedAt: 'sometime',
    completedAt: 'a different time',
    error: null,
  },
}

export const auditBoardMocks: {
  [key in 'empty' | 'unfinished' | 'finished']: IAuditBoard[]
} = {
  empty: [],
  unfinished: [
    {
      id: 'audit-board-1',
      name: 'Audit Board #01',
      signedOffAt: '',
      passphrase: 'happy rebel base',
      currentRoundStatus: {
        numSampledBallots: 30,
        numAuditedBallots: 0,
      },
    },
  ],
  finished: [
    {
      id: 'audit-board-1',
      name: 'Audit Board #01',
      signedOffAt: '',
      passphrase: 'happy rebel base',
      currentRoundStatus: {
        numSampledBallots: 30,
        numAuditedBallots: 30,
      },
    },
  ],
}
