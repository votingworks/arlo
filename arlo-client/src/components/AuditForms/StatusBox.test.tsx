import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import { statusStates } from './_mocks'
import StatusBox from './StatusBox'

describe('StatusBox', () => {
  it('renders a complete state', () => {
    const { container } = render(
      <StatusBox
        electionId="election-id"
        audit={statusStates[5]}
        launched
        started
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('renders the rounds in progress state', () => {
    const { container } = render(
      <StatusBox
        electionId="election-id"
        audit={statusStates[4]}
        launched
        started
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('downloads final report', async () => {
    window.open = jest.fn()
    const { findByText } = render(
      <StatusBox
        electionId="election-id"
        audit={statusStates[5]}
        launched
        started
      />
    )

    const button = await findByText('Download Audit Reports')
    fireEvent.click(button, { bubbles: true })

    expect(window.open).toBeCalledTimes(1)
    expect(window.open).toBeCalledWith('/election/election-id/audit/report')
  })
})
