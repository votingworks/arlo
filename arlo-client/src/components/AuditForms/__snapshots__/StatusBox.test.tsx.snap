// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatusBox renders a complete state 1`] = `
<div>
  <div
    class="bp3-callout sc-bwzfXH kMxkCT"
  >
    <h4
      class="bp3-heading"
    >
      The audit is complete.
    </h4>
    <div
      class="details"
    />
    <div
      class="wrapper"
    >
      <button
        class="bp3-button bp3-intent-success sc-bdVaJa gLVjYw"
        type="button"
      >
        <span
          class="bp3-button-text"
        >
          Download Audit Reports 
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`StatusBox renders the rounds in progress state 1`] = `
<div>
  <div
    class="bp3-callout sc-bwzfXH kMxkCT"
  >
    <h4
      class="bp3-heading"
    >
      Round 1 of the audit is in progress.
    </h4>
    <div
      class="details"
    />
    <div
      class="wrapper"
    >
      <button
        class="bp3-button bp3-disabled sc-bdVaJa gLVjYw"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <span
          class="bp3-button-text"
        >
          Download Audit Reports 
          (Incomplete)
        </span>
      </button>
    </div>
  </div>
</div>
`;
